### PR TITLE
Protect against infinite loops in IterativeOptimizer

### PR DESCRIPTION
**Pull request by mfussenegger**
This adds a similar mechanism to that of the simple `Optimizer` to
protect against infinite loops.

### CODE REVIEW

File: server/src/main/java/io/crate/planner/optimizer/iterative/IterativeOptimizer.java [hunk 1]
```diff
@@ -109,11 +110,26 @@ private boolean exploreNode(int group, Context context) {
...
+            int currentIt = i;
+            if (i % 100 == 0) {
+                long now = System.nanoTime();
+                long elapsed = now - start;
+                if (elapsed > maxPlanDurationInNs) {
+                    if (LOGGER.isTraceEnabled()) {
+                        LOGGER.trace("Spent {} in optimizer, aborting.", TimeValue.timeValueNanos(elapsed));
+                    }
+                    break;
+                }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/planner/optimizer/iterative/IterativeOptimizer.java:130]: Not sure if we want/need both a duration and max-iterations check. Wdyt?

**Comment by mkleen** [comment]
I have some second thoughts. I think i would prefer a mechanism which throws an error in the case of a infinite loop.

**Comment by mkleen** [comment]
In Presto there is a time limit https://github.com/prestodb/presto/commit/d359016e75978599bb99f2d67df2eacf3320ca5b

**Comment by mfussenegger** [comment]
> In Presto there is a time limit [prestodb/presto@d359016](https://github.com/prestodb/presto/commit/d359016e75978599bb99f2d67df2eacf3320ca5b)

I initially had one too but removed it again as it makes debugging more difficult, especially if the limit is low as you're likely to hit it.

**Comment by mkleen** [comment]
Another idea could be to record the rule applications by rule and maybe logical plan and look out for outliers.

**Comment by mfussenegger** [comment]
> Another idea could be to record the rule applications by rule and maybe logical plan and look out for outliers.

It would also make it more complex and add a bit of overhead. I think that goes against the point of protecting against logic bugs.

Regarding the timeout/duration limit: I think we have two different use-cases here:

- Prevent logic bugs from causing an infinite loop. You're right that we should probably trigger an assertion if it happens, so that we catch it in development, but don't fail queries in production which would run fine (although maybe less optimal)
- Max optimizer duration: I think this may become more interesting once we've join ordering optimizations where the permutations could explode the optimizer time. This isn't about logic bugs/infinite loops, but in general how much time the optimizer is allowed to take up.

**Comment by matriv** [comment]
I tend to agree with @mfussenegger.
Imho we should just add a loop counter for now but with an exception, why just a dev-only assertion? Isn't it better to stop execution in production after this 10k limit, instead of having the optimizer run "for ever"?

Indeed the timeout solution is an approach to stop the optimizer from checking a large amount of permutations, so imho we should implement it once we have this logic in the optimizer.

**Comment by mkleen** [comment]
I am ok with merging this change, but we should keep an eye on this.

**Pull request merged**